File: Atcoder/abc396_f_Rotated_Inversions.cpp
#include <bits/stdc++.h>
using ll = long long;
using ull = unsigned long long;
using namespace std;
#define fi first
#define se second
#define pii pair<int, int>
#define pll pair<long long, long long>
#define endl '\n'
#define rep(i, a, b) for (int i = (a); i < (b); ++i)
#define rep1(i, a, b) for (int i = (a); i <= (b); ++i)
#define rrep(i, a, b) for (int i = (a); i >= (b); --i)
#define AC return 0;
//#define int long long
//#define int unsigned long long
constexpr int MOD = 998'244'353;
constexpr int MOD_P = MOD - 1;
constexpr int mod = 1e9+7;
constexpr int INF = 0x3f3f3f3f;
constexpr ll LINF = 0x3f3f3f3f'3f3f3f3fLL;
constexpr int dx[] = {-1, 1, 0, 0};
constexpr int dy[] = {0, 0, -1, 1};
constexpr char dc[] = {'U', 'D', 'L', 'R'};

const double PI = acos(-1.0);
const int MAXN = 2'00'005;

void init() {

}

void solve() {
    int N, M; cin >> N >> M;
    vector<int> A(N);
    for (int& x : A) cin >> x;
    vector<ll> bit(M + 1, 0);
    auto add = [&](int pos, ll val) -> void {
        for (int i = pos + 1; i <= M; i += i & -i)
            bit[i] += val;
    };
    auto prefix = [&](int pos) -> ll {
        ll res = 0;
        for (int i = pos + 1; i > 0; i -= i & -i)
            res += bit[i];
        return res;
    };
    ll inv0 = 0;
    for (int x : A) {
        inv0 += prefix(M - 1) - prefix(x);
        add(x, 1);
    }
    vector<int> cnt(M), seen(M);
    for (int x : A) cnt[x] ++;
    vector<ll> delta(M);
    rep(p, 0, N) {
        int a = A[p];
        ll pd = p - seen[a];
        ll sd = (N - 1LL - p) - (cnt[a] - 1 - seen[a]);
        int k = M - 1 - a;
        delta[k] += pd - sd;
        seen[a] ++;
    }
    vector<ll> ans(M);
    ll cur = inv0;
    rep(k, 0, M) {
        ans[k] = cur;
        cur += delta[k];
    }
    rep(k, 0, M) cout << ans[k] << endl;
}

signed main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    init();
    int T = 1;
    //cin >> T;
    cout << fixed << setprecision(15);
    while (T--) {
        solve();
        if (T) {
            cout << '\n';
        }
    }
    AC
}

File: Atcoder/abc396_f_Rotated_Inversions.py
import sys

def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    N, M = data[0], data[1]
    A = data[2 : 2 + N]
    bit = [0] * (M + 1)
    inv0 = 0
    for x in A:
        i = M
        tot = 0
        while i > 0:
            tot += bit[i]
            i -= i & -i
        i = x + 1
        le = 0
        while i > 0:
            le += bit[i]
            i -= i & -i
        inv0 += tot - le
        i = x + 1
        while i <= M:
            bit[i] += 1
            i += i & -i
    cnt = [0] * M
    for x in A: cnt[x] += 1
    seen = [0] * M
    delta = [0] * M
    for p in range(N):
        a = A[p]
        before_diff = p - seen[a]
        after_diff = (N - 1 - p) - (cnt[a] - 1 - seen[a])
        delta[M - 1 - a] += before_diff - after_diff
        seen[a] += 1
    ans = [0] * M
    cur = inv0
    for k in range(M):
        ans[k] = cur
        cur += delta[k]
    sys.stdout.write('\n'.join(map(str, ans)))

if __name__ == '__main__':
    main()